KAFKA-10164; Throttle Create Topic, Create Partition and Delete Topic Operations (KIP-599, Part II, Admin Changes)

dajac opened the pull request:
This PR introduces the admin client changes related to KIP-599.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

dajac committed:
Admin Client

dajac committed:
cleanup

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/admin/CreatePartitionsOptions.java:
```diff
@@ -30,6 +30,7 @@
 public class CreatePartitionsOptions extends AbstractOptions<CreatePartitionsOptions> {
 
     private boolean validateOnly = false;
+    private boolean retryQuotaViolatedException = true;
```
Since the exception being retried is called ThrottlingQuotaExceededException, maybe this should say `retryOnQuotaViolation` or something like that without referring to the exception?

dajac replied:
`retryOnQuotaViolation` sounds good to me.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/admin/CreatePartitionsOptions.java:
```diff
@@ -48,4 +49,21 @@ public CreatePartitionsOptions validateOnly(boolean validateOnly) {
\ No newline at end of file
+
+    /**
+     * Set the retry QuotaViolatedException to indicate whether QuotaViolatedException
```
Several comments refer to QuotaViolatedException which is not an actual exception

rajinisivaram reviewed (commented):
@dajac Thanks for the PR, looks good, left a couple of minor comments. For the retry case, we are relying on the client retrying after the throttle time and stopping retry after request timeout. It will be good to test these.

dajac committed:
Address comments

dajac commented:
@rajinisivaram Thanks for the review. I have updated the PR to address your feedback. Regarding the test case that you suggested, I have added test cases but I am 100% sure that they cover what you had in mind. Please, let me know if I have misunderstood your suggestion.

rajinisivaram reviewed (approved):
@dajac Thanks for the updates, LGTM. 
I was looking for two tests, you added the one for request timeout. The second was to verify that the client mutes itself for throttle timeout (that is the retry backoff). But actually since we are relying on existing functionality for that, we don't need another test.

rajinisivaram commented:
ok to test
